### PR TITLE
Query level0 from UI

### PR DESCRIPTION
**Pull request by kdp-cloud**
This pull request is for the implementation of querying Level 0 elements from the User Interface.

### CODE REVIEW

File: server/app/crud.py [hunk 1]
```diff
@@ -155,5 +155,10 @@ def add_vector(
         return new_vector
 
 
-def get_vectors_for_user(database: Session, user: schemas.User) -> List[model.Vector]:
-    return database.query(model.Vector).filter(model.Vector.users.any(id=user.id)).all()
+def get_level0_for_user(database: Session, user: schemas.User) -> List[model.Vector]:
+    "Query all vectors from the database."
```
Comment (MaybeJustJames) [inline server/app/crud.py:159]: This docstring doesn't really reflect reality -> "... that a given user has access to."

File: server/app/vectors.py [hunk 2]
```diff
@@ -1,13 +1,25 @@
...
+    database: Session = Depends(deps.get_db),
+    current_user: schemas.User = Depends(deps.get_current_user),
+) -> List[Vector]:
+    v_list = crud.get_level0_for_user(database=database, user=current_user)
+    print("#" * 80)
+    print("Returning these vectors:")
+    [print(v) for v in v_list]
+    print("#" * 80)
+
+    return v_list
```
Comment (MaybeJustJames) [inline server/app/vectors.py:25]:
```suggestion
    return crud.get_level0_for_user(database=database, user=current_user)
```

File: client/src/Main.elm [hunk 3]
```diff
@@ -6,6 +6,7 @@ import Browser exposing (Document)
 import Browser.Dom exposing (Error(..))
 import Browser.Navigation as Nav
 import Color
+import Debug exposing (todo)
```
Comment (MaybeJustJames) [inline client/src/Main.elm]: Remove the import

File: client/src/Main.elm [hunk 4]
```diff
@@ -16,10 +17,11 @@ import Element.Region
 import File exposing (File)
 import File.Select as Select
 import Html exposing (Html, a)
-import Html.Attributes as HA
+import Html.Attributes as HA exposing (required)
```
Comment (MaybeJustJames) [inline client/src/Main.elm]: Why not `HA.required`?

File: client/src/Main.elm [hunk 5]
```diff
@@ -884,23 +987,23 @@ pieLabel slice ( label, _ ) =
         [ TypedSvg.Core.text label ]
```
Comment (MaybeJustJames) [inline client/src/Main.elm:986]: I don't think you really intended to use SVG here?

File: client/src/Main.elm [hunk 6]
```diff
@@ -30,7 +32,7 @@ import Task
 import TypedSvg exposing (g, svg, text_)
 import TypedSvg.Attributes exposing (dy, stroke, textAnchor, transform, viewBox)
 import TypedSvg.Core exposing (Svg)
-import TypedSvg.Types exposing (AnchorAlignment(..), Paint(..), Transform(..), em)
+import TypedSvg.Types exposing (AnchorAlignment(..), Paint(..), Transform(..), YesNo, em)
```
Comment (MaybeJustJames) [inline client/src/Main.elm]: You're not using SVG. Remove

File: client/src/Main.elm [hunk 7]
```diff
@@ -1189,16 +1292,16 @@ insertTable model =
                     , { header = none
                       , width = fillPortion 5
                       , view =
-                            \insert ->
-                                Input.button [ Font.color color.blue, Font.bold, Font.underline ] { onPress = Just (AppendInsert insert), label = Element.text insert.name }
+                            \level0 ->
+                                Input.button [ Font.color color.blue, Font.bold, Font.underline ] { onPress = Just (AppendInsert level0), label = Element.text level0.name }
```
Comment (MaybeJustJames) [inline client/src/Main.elm:1295]: This line is too long. You can add a newline somewhere and `elm-format` will help.

File: client/src/Main.elm [hunk 8]
```diff
@@ -1382,18 +1485,18 @@ update msg model =
             ( { model | description = newDescription }, Cmd.none )
 
         AppendInsert newInsert ->
-            if not (List.member newInsert.overhang (List.map .overhang model.selectedInserts)) then
+            if not (List.member newInsert.bsa1_overhang (List.map .bsa1_overhang model.selectedInserts)) then
                 ( { model
                     | selectedInserts = List.append model.selectedInserts [ newInsert ]
-                    , constructLength = List.sum (model.selectedBackbone.length :: newInsert.length :: List.map .length model.selectedInserts)
+                    , constructLength = List.sum (model.selectedBackbone.length :: String.length newInsert.sequence :: List.map (.sequence >> String.length) model.selectedInserts)
```
Comment (MaybeJustJames) [inline client/src/Main.elm:1490]: Line is too long

File: client/src/Main.elm [hunk 9]
```diff
@@ -1382,18 +1485,18 @@ update msg model =
...
                   }
                 , Cmd.none
                 )
 
             else
                 ( { model
-                    | selectedInserts = newInsert :: List.filter (\insert -> not (insert.overhang == newInsert.overhang)) model.selectedInserts
-                    , constructLength = List.sum (model.selectedBackbone.length :: List.map .length (newInsert :: List.filter (\insert -> not (insert.overhang == newInsert.overhang)) model.selectedInserts))
+                    | selectedInserts = newInsert :: List.filter (\level0 -> not (level0.bsa1_overhang == newInsert.bsa1_overhang)) model.selectedInserts
+                    , constructLength = List.sum (model.selectedBackbone.length :: List.map (.sequence >> String.length) (newInsert :: List.filter (\level0 -> not (level0.bsa1_overhang == newInsert.bsa1_overhang)) model.selectedInserts))
```
Comment (MaybeJustJames) [inline client/src/Main.elm:1498]: Lines are too long

File: client/src/Main.elm [hunk 10]
```diff
@@ -1511,6 +1611,15 @@ update msg model =
...
             ( { model | level0GenbankContent = Just content }, Cmd.none )
 
+        Level0Received (Ok level0s) ->
+            ( { model | insertList = level0s }, Cmd.none )
+
+        Level0Received (Err err) ->
+            ( model, Cmd.none )
+
+        RequestAllLevel0 ->
+            ( model, authenticatedGet (Maybe.withDefault "" model.token) "http://localhost:8000/vectors/level0" Level0Received (Decode.list level0Decoder) )
```
Comment (MaybeJustJames) [inline client/src/Main.elm:1620]: Line is too long.

I would be inclined to do something different is model.token is empty. E.g. display an error

```suggestion
            case model.token of
              Just token -> -- blah blah
              Nothing -> -- show error
```

File: client/src/Main.elm (hunk 3 above)
Comment (kdp-cloud) [reply client/src/Main.elm]: Already took care of that.

Comment (kdp-cloud) [reply client/src/Main.elm]: Took care of that in latest commit.

Comment (MaybeJustJames) [reply client/src/Main.elm]: I did mark it as resolved ;)

File: client/src/Main.elm (hunk 4 above)
Comment (kdp-cloud) [reply client/src/Main.elm]: This was a suggestion from the review test.
I can't find any occurence of HA. required.

File: client/src/Main.elm (hunk 6 above)
Comment (kdp-cloud) [reply client/src/Main.elm]: I use Svg for the visual representation.
If I remove one of these import, the code won't compile.

File: client/src/Main.elm (hunk 7 above)
Comment (kdp-cloud) [reply client/src/Main.elm:1295]: Done!

File: client/src/Main.elm (hunk 9 above)
Comment (kdp-cloud) [reply client/src/Main.elm:1498]: Done!

File: client/src/Main.elm (hunk 6 above)
Comment (MaybeJustJames) [reply client/src/Main.elm]: [This](https://github.com/vibbits/GGW-Elm/pull/18#discussion_r795492666) seems to be the only place you're using it?

File: client/src/Main.elm (hunk 8 above)
Comment (kdp-cloud) [reply client/src/Main.elm:1490]: Done!

File: client/src/Main.elm (hunk 10 above)
Comment (kdp-cloud) [reply client/src/Main.elm:1620]: OK, Here's my proposition:

```
RequestAllLevel0 ->
            case model.token of
                Just token ->
                    ( model
                    , authenticatedGet (Maybe.withDefault "" model.token)
                        "http://localhost:8000/vectors/level0"
                        Level0Received
                        (Decode.list level0Decoder)
                    )

                Nothing ->
                    ( { model
                        | error =
                            Just <|
                                showHttpError (BadBody "Permission denied! Token is empty!")
                      }
                    , Cmd.none
                    )
```

File: server/app/crud.py (hunk 1 above)
Comment (kdp-cloud) [reply server/app/crud.py:159]: You're right! I've updated the docstring.

File: server/app/vectors.py (hunk 2 above)
Comment (kdp-cloud) [reply server/app/vectors.py:25]: I'll leave the console prints out.
Is there a good / better way to log all activity?

File: client/src/Main.elm (hunk 5 above)
Comment (kdp-cloud) [reply client/src/Main.elm:986]: I use SVG for the visual representation of the construct.

File: client/src/Main.elm [hunk 11]
```diff
@@ -1511,6 +1615,30 @@ update msg model =
...
+
+        Level0Received (Err _) ->
+            ( model, Cmd.none )
+
+        RequestAllLevel0 ->
+            case model.token of
+                Just token ->
+                    ( model
+                    , authenticatedGet token
+                        "http://localhost:8000/vectors/level0"
```
Comment (MaybeJustJames) [inline client/src/Main.elm:1629]: Can you make an issue: Configure API server address using `flags`

**Review by MaybeJustJames — APPROVED** [review]

**Pull request merged**